letsencrypt/client/plugins/nginx/configurator.py

```diff
@@ -0,0 +1,554 @@
+        try:
+            return self._enhance_func[enhancement](
+                self.choose_vhost(domain), options)
+        except (KeyError, ValueError):
+            raise errors.LetsEncryptConfiguratorError(
+                "Unsupported enhancement: {0}".format(enhancement))
+        except errors.LetsEncryptConfiguratorError:
+            logging.warn("Failed %s for %s", enhancement, domain)
+
+    ######################################
+    # Nginx server management (IInstaller)
+    ######################################
+    def restart(self):
+        """Restarts nginx server.
+
+        :returns: Success
+        :rtype: bool
+
+        """
+        return nginx_restart(self.config.nginx_ctl)
+
+    def config_test(self):  # pylint: disable=no-self-use
+        """Check the configuration of Nginx for errors.
+
+        :returns: Success
+        :rtype: bool
+
+        """
+        try:
+            proc = subprocess.Popen(
+                [self.config.nginx_ctl, "-t"],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE)
+            stdout, stderr = proc.communicate()
+        except (OSError, ValueError):
+            logging.fatal("Unable to run nginx config test")
+            sys.exit(1)
+
+        if proc.returncode != 0:
+            # Enter recovery routine...
+            logging.error("Config test failed")
+            logging.error(stdout)
+            logging.error(stderr)
+            return False
+
+        return True
+
+    def _verify_setup(self):
+        """Verify the setup to ensure safe operating environment.
+
+        Make sure that files/directories are setup with appropriate permissions
+        Aim for defensive coding... make sure all input files
+        have permissions of root.
+
+        """
+        uid = os.geteuid()
+        le_util.make_or_verify_dir(self.config.work_dir, 0o755, uid)
+        le_util.make_or_verify_dir(self.config.backup_dir, 0o755, uid)
+        le_util.make_or_verify_dir(self.config.config_dir, 0o755, uid)
+
+    def get_version(self):
+        """Return version of Nginx Server.
+
+        Version is returned as tuple. (ie. 2.4.7 = (2, 4, 7))
+
+        :returns: version
+        :rtype: tuple
+
+        :raises errors.LetsEncryptConfiguratorError:
+            Unable to find Nginx version or version is unsupported
+
+        """
+        try:
+            proc = subprocess.Popen(
+                [self.config.nginx_ctl, "-V"],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE)
+            text = proc.communicate()[1]  # nginx prints output to stderr
+        except (OSError, ValueError):
+            raise errors.LetsEncryptConfiguratorError(
+                "Unable to run %s -V" % self.config.nginx_ctl)
+
+        version_regex = re.compile(r"nginx/([0-9\.]*)", re.IGNORECASE)
+        version_matches = version_regex.findall(text)
+
+        sni_regex = re.compile(r"TLS SNI support enabled", re.IGNORECASE)
```

Comment:
These checks are great!  I need to make a note to add some more checks to Apache.